### PR TITLE
feat: add substrait.proto convenience module and document it

### PR DESCRIPTION
**Pull request by amol-**
Add a ``substrait.proto`` module that gives access to the Substrait protocol classes 
removing the need to navigate the hierarchy automatically generated by protobuf. 
It also provides access to the modules without the ``_pb2`` suffix 
which is an implementation detail of the protobuf version used.

Provides examples on how to generate and read back Substrait plans 
using the substrait-python module itself.

### CODE REVIEW

File: tests/test_proto.py [hunk 1]
```diff
@@ -12,3 +12,21 @@ def test_imports():
     from substrait.gen.proto.type_expressions_pb2 import DerivationExpression
     from substrait.gen.proto.type_pb2 import Type
     from substrait.gen.proto.extensions.extensions_pb2 import SimpleExtensionURI
+
```
Comment (danepitkin) [inline tests/test_proto.py:15]: We can go ahead and delete the test_imports() function since the new test indirectly tests these imports.

**Review by danepitkin — APPROVED** [review]
LGTM! Thanks @amol- this is already much cleaner for the user :)

File: tests/test_proto.py [hunk 2]
```diff
@@ -12,3 +12,21 @@ def test_imports():
...
+
+def test_proto_proxy_module():
+    """Test that protocol classes are made available in substrait.proto"""
+    import substrait.proto
+
+    assert {"Plan", "Type", "NamedStruct", "RelRoot"} & set(dir(substrait.proto))
```
Comment (gforsyth) [inline tests/test_proto.py:21]:
```suggestion
    assert {"Plan", "Type", "NamedStruct", "RelRoot"} <= set(dir(substrait.proto))
```

I'm assuming here that the goal is checking that these know values are a subset of the contents of `substrait.proto`?

File: tests/test_proto.py [hunk 3]
```diff
@@ -12,3 +12,21 @@ def test_imports():
...
+        "algebra",
+        "capabilities",
+        "extensions",
+        "extended_expression",
+        "function",
+        "parameterized_types",
+        "plan",
+        "type_expressions",
+        "type",
+    } & set(dir(substrait.proto))
```
Comment (gforsyth) [inline tests/test_proto.py:32]:
```suggestion
    } <= set(dir(substrait.proto))
```

**Review by gforsyth — COMMENTED** [review]
looks good overall -- two small suggestions for the tests.

**Comment by EpsilonPrime** [comment]
So the purpose of this PR is to get the protos exposed with the simpler package name without changing the fact that they are registered with the existing one?

One reason the current package name is used is to prevent conflicts when the C++ version of the protos is also loaded under ::substrait::proto.  We should probably document this desire along with the fact that we aren't simply changing what package we generate the protos in.

File: tests/test_proto.py (hunk 2 above)
Comment (amol-) [reply tests/test_proto.py:21]: Ops, good catch, it would have passed even if only one existed

File: tests/test_proto.py (hunk 1 above)
Comment (amol-) [reply tests/test_proto.py:15]: I think there is a value in keeping this test to ensure backward compatibility.
`substrait.proto` might continue to work but import the classes from modules with different names than the current ones. 

In such case we would be breaking backward compatibility for anyone directly accessing the classes from the `_pb2` modules and that's something we want to catch and be aware of.

**Comment by amol-** [comment]
> So the purpose of this PR is to get the protos exposed with the simpler package name without changing the fact that they are registered with the existing one?
> 

Correct, I wanted to retain backward compatibility for anyone importing the classes from the autogenerated modules, but also make the automatic generation an implementation detail.

> One reason the current package name is used is to prevent conflicts when the C++ version of the protos is also loaded under ::substrait::proto. We should probably document this desire along with the fact that we aren't simply changing what package we generate the protos in.

Can you elaborate? How does the C++ library namespace influence the Python namespace?

**Comment by EpsilonPrime** [comment]
Depending on your options you can actually use C++ protos for the Python under the hood.  But ignoring that you can still end up with two definitions for protobufs which will collide and cause an exception when the module loads.  If they are in different namespaces that collision won't happen (you can pretend to have two different sets of protos).

On a side note, I have been having issues with the current protos (lint and IDE) because the attributes aren't available for the package until runtime.  The solution I used for the spark connect protos was to use mypy-protobuf (add --mypy_out to the protoc call in gen_proto.sh) which created pyi files which also would need to be checked in.  Would this approach work here?

**Review by EpsilonPrime — APPROVED** [review]

**Review by gforsyth — APPROVED** [review]
thanks @amol!

**Pull request merged**